iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,327 @@
+use std::borrow::BorrowMut;
+use std::cmp::Ordering;
+use std::future::Future;
+use std::io::{Error, ErrorKind};
+use std::net::SocketAddr;
+use std::ops::{Deref, DerefMut};
+use std::sync::Arc;
+
+use async_stream::stream;
+use futures::Stream;
+use tokio::io;
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWriteExt};
+use tokio::net::tcp::{OwnedReadHalf, OwnedWriteHalf};
+use tokio::net::TcpStream;
+use tokio::sync::Mutex;
+use tokio::time::Instant;
+
+use iroha_actor::{Actor, Context, Handler, MessageResponse, Addr, ContextHandler};
+use iroha_crypto::{Hash};
+use iroha_data_model::prelude::PeerId;
+use iroha_logger::{debug, error, info, trace, warn};
+
+use crate::message::{Message, MessageResult};
+use crate::network::{SendPeerMsg, ReceivedPeerMsg, Connect};
+use std::pin::Pin;
+use crate::Network;
+use rand::{Rng, RngCore};
+use crate::chacha::Chacha;
+use x25519_dalek::{PublicKey, StaticSecret};
+
+const MAX_MESSAGE_LENGTH: usize = 2 * 1024 * 1024;
+
+pub struct Peer {
+    id: PeerId,
+    read: Option<OwnedReadHalf>,
+    write: Option<OwnedWriteHalf>,
```

Comment:
i don't see places where you don't have WriteHalf

Reply:
The whole `TcpStream` (and it's parts) is missing when we create this `Peer` to connect to some other node. And after the connection we fill those fields.

Reply:
You call Peer::new only once and you actually have stream there :)

Reply:
Yep, didn't finish that `connect` code.